pkg/server/plugin/nodeattestor/aws/client.go

```diff
@@ -20,49 +21,55 @@ type Client interface {
 }
 
 type clientsCache struct {
-	mu        sync.RWMutex
+	mtx       sync.RWMutex
 	config    *SessionConfig
-	clients   map[string]Client
+	clients   map[string]*cacheEntry
 	newClient newClientCallback
 }
 
-type newClientCallback func(config *SessionConfig, region string, asssumeRoleARN string) (Client, error)
+type cacheEntry struct {
+	lock   chan struct{}
+	client Client
+}
+
+type newClientCallback func(ctx context.Context, config *SessionConfig, region string, asssumeRoleARN string) (Client, error)
 
 func newClientsCache(newClient newClientCallback) *clientsCache {
 	return &clientsCache{
-		clients:   make(map[string]Client),
+		clients:   make(map[string]*cacheEntry),
 		newClient: newClient,
 	}
 }
 
 func (cc *clientsCache) configure(config SessionConfig) {
-	cc.mu.Lock()
-	cc.clients = make(map[string]Client)
+	cc.mtx.Lock()
+	cc.clients = make(map[string]*cacheEntry)
 	cc.config = &config
-	cc.mu.Unlock()
+	cc.mtx.Unlock()
 }
 
-func (cc *clientsCache) getClient(region, accountID string) (Client, error) {
-	// do an initial check to see if p client for this region already exists
+func (cc *clientsCache) getClient(ctx context.Context, region, accountID string) (Client, error) {
+	// Do an initial check to see if p client for this region already exists
 	cacheKey := accountID + "@" + region
 
-	cc.mu.RLock()
-	client, ok := cc.clients[region]
-	cc.mu.RUnlock()
-	if ok {
-		return client, nil
+	// Grab (or create) the cache for the region
+	r := cc.getCachedClient(cacheKey)
+
+	// Obtain the "lock" to the region cache
+	select {
+	case <-ctx.Done():
```

Comment:
Just confirming, is this change in the synchronization of concurrent map access driven by us trying to respect context cancellations during execution of this method?